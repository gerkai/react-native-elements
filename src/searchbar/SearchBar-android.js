import PropTypes from 'prop-types';
import React, { Component } from 'react';
import {
  Dimensions,
  StyleSheet,
  View,
  ActivityIndicator,
  Text,
} from 'react-native';

import ViewPropTypes from '../config/ViewPropTypes';
import nodeType from '../helpers/nodeType';
import Input from '../input/Input';
import Icon from '../icons/Icon';
import renderNode from '../helpers/renderNode';

const SCREEN_WIDTH = Dimensions.get('window').width;
const ANDROID_GRAY = 'rgba(0, 0, 0, 0.54)';
const defaultSearchIcon = {
  type: 'material-community',
  size: 25,
  color: ANDROID_GRAY,
  name: 'magnify',
};

class SearchBar extends Component {
  focus = () => {
    this.input.focus();
  };

  blur = () => {
    this.input.blur();
  };

  clear = () => {
    this.input.clear();
    this.onChangeText('');
    this.props.onClear();
  };

  cancel = () => {
    this.blur();
    this.props.onCancel();
  };

  onFocus = () => {
    this.props.onFocus();
    this.setState({ hasFocus: true });
  };

  onBlur = () => {
    this.props.onBlur();
    this.setState({ hasFocus: false });
  };

  onChangeText = text => {
    this.props.onChangeText(text);
    this.setState({ isEmpty: text === '' });
  };

  constructor(props) {
    super(props);
    this.state = {
      hasFocus: false,
      isEmpty: true,
    };
  }

  render() {
    const {
      clearIcon,
      containerStyle,
      leftIconContainerStyle,
      rightIconContainerStyle,
      inputContainerStyle,
      inputStyle,
      searchIcon,
      cancelIcon,
      showLoading,
      loadingProps,
      ...attributes
    } = this.props;
    const { hasFocus, isEmpty } = this.state;
    const { style: loadingStyle, ...otherLoadingProps } = loadingProps;

    const defaultCancelIcon = {
      type: 'material-community',
      size: 25,
      color: ANDROID_GRAY,
      name: 'arrow-left',
      onPress: this.cancel,
    };
    const defaultClearIcon = {
      type: 'material-community',
      name: 'close',
      size: 25,
      color: ANDROID_GRAY,
      onPress: this.clear,
    };
    return (
      <View style={[styles.container, containerStyle]}>
        <Input
          {...attributes}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onChangeText={this.onChangeText}
          ref={input => (this.input = input)}
          inputStyle={[styles.input, inputStyle]}
          inputContainerStyle={[styles.inputContainer, inputContainerStyle]}
          leftIcon={
            hasFocus
              ? renderNode(Icon, cancelIcon, defaultCancelIcon)
              : renderNode(Icon, searchIcon, defaultSearchIcon)
          }
          leftIconContainerStyle={[
            styles.leftIconContainerStyle,
            leftIconContainerStyle,
          ]}
          rightIcon={
            <View style={{ flexDirection: 'row' }}>
              {showLoading && (
                <ActivityIndicator
                  style={[{ marginRight: 5 }, loadingStyle]}
                  {...otherLoadingProps}
                />
              )}
              {!isEmpty && renderNode(Icon, clearIcon, defaultClearIcon)}
            </View>
          }
          rightIconContainerStyle={[
            styles.rightIconContainerStyle,
            rightIconContainerStyle,
          ]}
        />
      </View>
    );
  }
}

SearchBar.propTypes = {
  clearIcon: nodeType,
  searchIcon: nodeType,
  cancelIcon: nodeType,
  loadingProps: PropTypes.object,
  showLoading: PropTypes.bool,
  containerStyle: ViewPropTypes.style,
  leftIconContainerStyle: ViewPropTypes.style,
  rightIconContainerStyle: ViewPropTypes.style,
  inputContainerStyle: ViewPropTypes.style,
  inputStyle: Text.propTypes.style,
  onClear: PropTypes.func,
  onCancel: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  onChangeText: PropTypes.func,
};

SearchBar.defaultProps = {
  loadingProps: {},
  showLoading: false,
  onClear: () => null,
  onCancel: () => null,
  onFocus: () => null,
  onBlur: () => null,
  onChangeText: () => null,
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    width: SCREEN_WIDTH,
    paddingTop: 8,
    paddingBottom: 8,
  },
  input: {
    marginLeft: 24,
    marginRight: 8,
  },
  inputContainer: {
    borderBottomWidth: 0,
    width: SCREEN_WIDTH,
  },
  rightIconContainerStyle: {
    marginRight: 8,
  },
  leftIconContainerStyle: {
    marginLeft: 8,
  },
});

export default SearchBar;
