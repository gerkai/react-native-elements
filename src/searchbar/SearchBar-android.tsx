import React, { Component } from 'react';
import {
  StyleSheet,
  View,
  ActivityIndicator,
  Keyboard,
  TextInput,
} from 'react-native';
import { renderNode } from '../helpers';
import Input, { InputProps } from '../input/Input';
import Icon, { IconNode } from '../icons/Icon';
import { SearchBarBaseProps } from './SearchBar';
import { ThemeProps } from '../config';

const defaultSearchIcon = (theme) => ({
  type: 'material',
  size: 25,
  color: theme.colors.platform.android.grey,
  name: 'search',
});

const defaultCancelIcon = (theme) => ({
  type: 'material',
  size: 25,
  color: theme.colors.platform.android.grey,
  name: 'arrow-back',
});

const defaultClearIcon = (theme) => ({
  type: 'material',
  size: 25,
  color: theme.colors.platform.android.grey,
  name: 'clear',
});

export type SearchBarAndroidProps = InputProps &
  SearchBarBaseProps &
  typeof SearchBar.defaultProps & {
    cancelIcon?: IconNode;
  };

type SearchBarState = {
  hasFocus: boolean;
  isEmpty: boolean;
};

class SearchBar extends Component<
  SearchBarAndroidProps & Partial<ThemeProps<SearchBarAndroidProps>>,
  SearchBarState
> {
  input!: TextInput;
  static defaultProps = {
    onClear: () => null,
    onCancel: () => null,
    onFocus: () => null,
    onBlur: () => null,
    onChangeText: () => null,
  };

  focus = () => {
    this.input.focus();
  };

  blur = () => {
    this.input.blur();
  };

  clear = () => {
    this.input.clear();
    this.onChangeText('');
    this.props.onClear();
  };

  cancel = () => {
    this.blur();
    this.props.onCancel();
  };

  onFocus: InputProps['onFocus'] = (event) => {
    this.props.onFocus(event);
    this.setState({
      hasFocus: true,
      isEmpty: this.props.value === '',
    });
  };

  onBlur: InputProps['onBlur'] = (event) => {
    this.props.onBlur(event);
    this.setState({ hasFocus: false });
  };

  onChangeText = (text: string) => {
    this.props.onChangeText(text);
    this.setState({ isEmpty: text === '' });
  };

  constructor(props: SearchBarAndroidProps) {
    super(props);
    const { value = '' } = props;
    this.state = {
      hasFocus: false,
      isEmpty: value ? value === '' : true,
    };
    Keyboard.addListener('keyboardDidHide', this._keyboardDidHide);
  }

  _keyboardDidHide = () => {
    this.cancel();
  };

  componentWillUnmount() {
    Keyboard.removeListener('keyboardDidHide', this._keyboardDidHide);
  }

  render() {
    const {
      theme,
      clearIcon = { name: 'clear' },
      containerStyle,
      leftIconContainerStyle,
      rightIconContainerStyle,
      inputContainerStyle,
      inputStyle,
      searchIcon = { name: 'search' },
      cancelIcon = { name: 'arrow-back' },
      showLoading = false,
      loadingProps = {},
      ...attributes
    } = this.props;
    const { hasFocus, isEmpty } = this.state;
    const { style: loadingStyle, ...otherLoadingProps } = loadingProps;

    return (
      <View
        style={StyleSheet.flatten([
          {
            backgroundColor: theme.colors.white,
            paddingTop: 8,
            paddingBottom: 8,
          },
          containerStyle,
        ])}
      >
        <Input
          testID="searchInput"
          renderErrorMessage={false}
          {...attributes}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onChangeText={this.onChangeText}
          //@ts-ignore
          ref={(input: TextInput) => {
            this.input = input;
          }}
          containerStyle={{ paddingHorizontal: 0 }}
          inputStyle={StyleSheet.flatten([styles.input, inputStyle])}
          inputContainerStyle={StyleSheet.flatten([
            styles.inputContainer,
            inputContainerStyle,
          ])}
          leftIcon={
            hasFocus
              ? renderNode(Icon, cancelIcon, {
                  ...defaultCancelIcon(theme),
                  onPress: this.cancel,
                })
              : renderNode(Icon, searchIcon, defaultSearchIcon(theme))
          }
          leftIconContainerStyle={StyleSheet.flatten([
            styles.leftIconContainerStyle,
            leftIconContainerStyle,
          ])}
          rightIcon={
            <View style={{ flexDirection: 'row' }}>
              {showLoading && (
                <ActivityIndicator
                  key="loading"
                  style={StyleSheet.flatten([{ marginRight: 5 }, loadingStyle])}
                  {...otherLoadingProps}
                />
              )}
              {!isEmpty &&
                renderNode(Icon, clearIcon, {
                  ...defaultClearIcon(theme),
                  key: 'cancel',
                  onPress: this.clear,
                })}
            </View>
          }
          rightIconContainerStyle={StyleSheet.flatten([
            styles.rightIconContainerStyle,
            rightIconContainerStyle,
          ])}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  input: {
    marginLeft: 24,
    marginRight: 8,
  },
  inputContainer: {
    borderBottomWidth: 0,
    width: '100%',
  },
  rightIconContainerStyle: {
    marginRight: 8,
  },
  leftIconContainerStyle: {
    marginLeft: 8,
  },
});

export default SearchBar;
