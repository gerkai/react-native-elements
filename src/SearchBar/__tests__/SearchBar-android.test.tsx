import { SearchBarAndroid } from '../SearchBar-android';
import { commonTests } from './common';

describe('Android SearchBar component', () => {
  commonTests(SearchBarAndroid);
});
