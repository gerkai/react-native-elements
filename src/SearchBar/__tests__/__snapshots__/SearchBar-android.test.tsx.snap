// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Android SearchBar component should render without issues 1`] = `
<View
  style={
    Object {
      "backgroundColor": undefined,
      "paddingBottom": 8,
      "paddingTop": 8,
    }
  }
  testID="RNE__SearchBar-wrapper"
>
  <View
    style={
      Object {
        "paddingHorizontal": 0,
        "width": "100%",
      }
    }
    testID="RNE__Input__view-wrapper"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 0,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 40,
            "justifyContent": "center",
            "marginLeft": 8,
            "marginVertical": 4,
            "paddingRight": 4,
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            replaceTheme={[Function]}
            updateTheme={[Function]}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#242424",
                      "fontSize": 25,
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                    Object {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
                testID="iconIcon"
              >
                î¢¶
              </Text>
            </View>
          </View>
        </View>
      </View>
      <TextInput
        allowFontScaling={true}
        editable={true}
        onBlur={[Function]}
        onCancel={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="#86939e"
        rejectResponderTermination={true}
        style={
          Object {
            "color": "#242424",
            "flex": 1,
            "fontSize": 18,
            "marginLeft": 24,
            "marginRight": 8,
            "minHeight": 40,
          }
        }
        testID="RNE__SearchBar"
        underlineColorAndroid="transparent"
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 40,
            "justifyContent": "center",
            "marginRight": 8,
            "marginVertical": 4,
            "paddingRight": 4,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        />
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "#ff190c",
          "fontSize": 12,
          "height": 0,
          "margin": 0,
          "padding": 0,
        }
      }
    />
  </View>
</View>
`;
